### ISSUE TITLE
Draw tool does not snap to grid correctly

### ISSUE THREAD

**Issue by jmickle66666666** [sabresaurus/SabreCSG#219]
![sabredraw2](https://user-images.githubusercontent.com/7083717/51444609-e8d1aa80-1cf9-11e9-8033-da331a13bb72.gif)

When using the Draw tool, it is snapping to the correct grid size, but not the global position of the grid properly. The Vertex tool and other tools match up with the projected grid perfectly but the draw tool doesn't.

**Comment by damagefilter** [comment]
It's because the tool tries to account for angled surfaces to attach cleanly to them.
It would snap the position in local space rather than the global grid.

Snapping the hit point to the global grid will cause hovering brushes on planes that are not on the grid.

Supposedly, one could work out a check for the surface normal we're attaching to and if it's a straight cardinal direction  use a global snap, otherwise keep the local snap to prevent a brush from hovering.

**Comment by jmickle66666666** [comment]
The grid corners you can see on the gif are worldspace grid coordinates that exist on the plane, snapping to the global grid does not mean the points would float in space.

**Comment by jmickle66666666** [comment]
At the very least it would be incredibly useful to be able to choose to switch between plane-grid and global grid when using the draw tool. Using the draw tool on angled surfaces without drawing on the world grid makes the resulting geometry really unwieldy.

**Comment by damagefilter** [comment]
What I mean is, if we are to snap the vertex points when drawing to a global grid,
they will not attach to a plane that is not intersecting the grid, like here.
![image](https://user-images.githubusercontent.com/1213756/51472520-fd558780-1d79-11e9-9267-0c39511b12f0.png)

Of course, if that plane is intersecting the grid then there is no issue,
![image](https://user-images.githubusercontent.com/1213756/51472591-3988e800-1d7a-11e9-81ef-c178a4f3333e.png)

I suppose that's a bit too naive for snapping.
So I figured, perhaps we can find another approach.
Maybe only snap 2 components of the hitpoint vector.
Something along those lines ...

**Comment by jmickle66666666** [comment]
Yeah that's basically what i'm getting at. I understand the current snapping logic but I think it's less useful (in my experience) than the 2 axes system the projected grid represents

**Comment by jmickle66666666** [comment]
![sabrecgrid](https://user-images.githubusercontent.com/7083717/51786171-08852a80-2161-11e9-9b52-ea25f7e32abb.gif)
even on an axis-aligned surface the grid is offset

**Comment by damagefilter** [comment]
Yeah. I have noticed that too and fixed that bit for sure.

But the slopes ...
I had two versions.
One where it would be on-grid on slopes but would not attach to the surface.
(It would snap to points below or beneath the surfaces plane on odd angles, the usual))
One where it would attach to the surface but remain off-grid.

Can't seem to wrap my head around that part :C

**Comment by Henry00IS** [comment]
@jmickle66666666 is currently working on [adding concave shape support](https://github.com/jmickle66666666/SabreCSG/tree/draw-tool-concave) to the draw tool as well. You may wish to speak with her before you both end up accidentally fixing the same issue. 😁

**Comment by jmickle66666666** [comment]
i actually managed to sort out the grid alignment stuff. i should take that out into a new branch first before doing the concave stuff, since that's gonna be more difficult

**Comment by damagefilter** [comment]
Yes that's why I said the thing with the 2 brains, so we could perhaps look into that together.
Since @jmickle66666666 got a fix for it now though, I'm excited for how the solution looks like.
